### PR TITLE
fix: resolve distributed JIT checkpoint intermittent resume failure

### PR DESCRIPTION
**Pull request by efazal**
**What this PR does / why we need it**:

Fixes [RHOAIENG-45711](https://issues.redhat.com/browse/RHOAIENG-45711)

Resume training fails because incomplete JIT checkpoints appear complete due to missing sentinel markers. Root cause: race conditions in distributed training during pause/preemption.

Changes

  - Remove checkpoint thread entirely: Replaced daemon thread with direct checkpoint call from SIGTERM handler. If SIGTERM arrives during an optimizer step (where direct checkpoint would deadlock), checkpoint is deferred to the on_optimizer_step callback via a _checkpoint_deferred flag.
  - Fix step desynchronization: Added barrier before reading global_step to ensure all ranks checkpoint at the same training step (prevents checkpoint-29 vs checkpoint-30 splits in multi node training)
  - Fix incomplete marker handling: All ranks now write/remove their own per-rank sentinel markers instead of only global rank 0
  - Fix barrier warnings: Added device specification to all dist.barrier() calls to eliminate PyTorch warnings
  - Exit process after JIT checkpoint: Added sys.exit(0) in on_train_end after checkpoint and S3 upload complete to prevent user code (e.g. trainer.save_model()) from running redundantly during graceful shutdown.
  
  
image
image
image

## Summary by CodeRabbit

* **Bug Fixes**
  * Checkpointing now uses per-rank, per-node coordination to avoid cross-rank conflicts and incomplete saves.
  * Incomplete checkpoints are now detected and skipped to prevent partial restores.
  * Shutdown handling triggers synchronous checkpointing when safe and performs an early checkpoint if shutdown occurs during critical optimizer work to prevent deadlocks.
  * Synchronization tightened at init, save, download, and training start; CUDA-aware barriers avoid device-ID warnings.
  * Logs include node/local-rank context and clearer guidance for manual cleanup.

### CODE REVIEW

File: kubeflow/trainer/rhai/transformers.py [hunk 1]
```diff
@@ -265,12 +282,20 @@ def _sigterm_handler(self, signum, frame):
...
             self.checkpoint_thread = threading.Thread(
-                target=self._async_checkpoint, daemon=True, name="KubeflowJITCheckpoint"
+                target=self._async_checkpoint, daemon=False, name="KubeflowJITCheckpoint"
             )
             self.checkpoint_thread.start()
 
+            # Wait for checkpoint to complete before allowing process to terminate
+            _log("Waiting for checkpoint thread to complete...")
+            self.checkpoint_thread.join()
+            _log("Checkpoint thread completed")
```
[the remark this answers is not in the document]
Comment (efazal) [reply kubeflow/trainer/rhai/transformers.py]: Since the job runs on k8s, the kubelet always sends SIGTERM at the same time. However, in the case of a worker terminating abruptly due to node drain or worker pod crashing, regardless of this implementation, the other worker continue training until they hit a barrier and realize one of the workers is absent and eventually timeout after waiting for the pod. This is an existing limitation. Also, the reason for the barrier at the sigterm_handler level is to avoid creating imcomplete checkpoints, where the training cannot actually resume from it.

[the remark this answers is not in the document]
Comment (efazal) [reply kubeflow/trainer/rhai/transformers.py]: Yes, will add the timeout.

File: kubeflow/trainer/rhai/transformers.py [hunk 2]
```diff
@@ -233,6 +233,23 @@ def _log(message: str, args: Optional[object] = None) -> None:
...
         print(f"{_log_prefix} {message}", flush=True)
 
+    def _wait_for_all_ranks(operation: str) -> None:
+        """Barrier across ranks to synchronize distributed training."""
+        if not dist.is_available() or not dist.is_initialized():
+            return
+        try:
+            # Specify device to avoid "guessing device ID" warning
+            if torch.cuda.is_available():
+                dist.barrier(device_ids=[torch.cuda.current_device()])
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py:244]: question: what happens if there's multiple devices on this node?

File: kubeflow/trainer/rhai/transformers.py [hunk 3]
```diff
@@ -265,12 +282,26 @@ def _sigterm_handler(self, signum, frame):
...
+            # Wait for checkpoint to complete before allowing process to terminate
+            _log("Waiting for checkpoint thread to complete...")
+            self.checkpoint_thread.join(timeout=300)
+            if self.checkpoint_thread.is_alive():
+                _log(
+                    "Warning: Checkpoint thread still running after 5 mins timeout. "
+                    "Process will terminate, checkpoint may be incomplete."
+                )
+            else:
+                _log("Checkpoint thread completed")
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: do we need the thread? And don't we need to be able to wait longer than 5 mins (once your pytorch PR is merged)?

```
self._async_checkpoint()
```

File: kubeflow/trainer/rhai/transformers.py [hunk 4]
```diff
@@ -1049,7 +1070,11 @@ def _patched_train(resume_from_checkpoint=None, **train_kwargs):
                         args=training_args,
                     )
                     try:
-                        dist.barrier()
+                        # Specify device to avoid "guessing device ID" warning
+                        if torch.cuda.is_available():
+                            dist.barrier(device_ids=[torch.cuda.current_device()])
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py:1093]: same question as above - what happens if there's multiple devices on this node?

File: kubeflow/trainer/rhai/transformers.py (hunk 2 above)
Comment (efazal) [reply kubeflow/trainer/rhai/transformers.py:244]: Each rank invokes `torch.cuda.current_device()` which will return the GPU ID it's assigned.

File: kubeflow/trainer/rhai/transformers.py (hunk 3 above)
Comment (efazal) [reply kubeflow/trainer/rhai/transformers.py]: Yeah we need the thread to not block the main thread so when JIt is triggered and the training is in optimizer step, we cannot checkpoint and we have to wait. If we remove thread, it will cause a deadlock. And other activities like progression tracking can continue in the background, s3 upload progress continue to be reported.

Sorry, I just realized I need to move the join() to some place else. I will update it. Good question about the timeout, I thought about the same. Should I set it based on TORCH_ELASTIC_SHUTDOWN_TIMEOUT and default to 30s otherwise (k8s default)?

Comment (efazal) [reply kubeflow/trainer/rhai/transformers.py]: @robert-bell, the moment we move join, the race condition issue returns. So the real fix is not even barrier synchronization, but rather blocking GPU during checkpoint. We can make the JIT checkpointing synchronous, by triggering JIT checkpoint from multiple callbacks like on `on_step_end()` , `on_step_begin`, and `on_epoch_end`, which is safe because we are certain that the traning is not in optimizer stage. This was it's more robust, but at the cost of being not truly JIT checkpoint. But with the timeout fix in Pytorch on the horizon, I think this isn't a big deal if the JIT checkpointing begins slightly late? WDYT?

Comment (robert-bell) [reply kubeflow/trainer/rhai/transformers.py]: That makes sense why it needs to be in a separate thread, but i'm still a bit confused.

Is the problem we're trying to avoid that the sigterm could be received during the optimization step? We can't take a checkpoint then because the gradients are being updated.

If the sigterm is NOT received during the optimisation step, can just directly ask transformers to do the checkpoint immediately for us?

I'm wondering - can we simplify this logic and get the handler to look like this:
- if sigterm received inside optimisation loop
  - set flag on CheckpointManager indicating jit checkpoint is requested
  - return immediately.
  - check the flag in on_optimizer_step, and if set take the checkpoint
- otherwise (sigterm not in optimisation loop)
  - use the current approach

Comment (robert-bell) [reply kubeflow/trainer/rhai/transformers.py]: Also - how is the JIT checkpoint uploaded to s3?

Comment (efazal) [reply kubeflow/trainer/rhai/transformers.py]: Thats a neat idea. I shall do that. 

After we invoke `_save_checkpoint()` the `on_save` callback is triggered, which will upload to S3 as it does for periodic checkpoint. 
But it seems like, `on_save` doesn't get triggered since we directly call it save checkpoint, and not something that happens within the transformer training loop. Good question. I can invoke on save directly right after save cehckpoints. Thanks for the inputs. It's very useful.

Comment (robert-bell) [reply kubeflow/trainer/rhai/transformers.py]: Np - glad I could help :) 

This needs some good manual testing btw to make sure that it works.

In particular, can you do a manual test that the checkpoint during the optimise step actually works? We might be able to 
- artificially inflate the optimise step time using a monkeypatch,
- stick some print statements in the monkeypatch before/after the actual optimise call,
- then manually trigger the sigterm and make sure the checkpoint is taken.

File: kubeflow/trainer/rhai/transformers.py [hunk 5]
```diff
@@ -928,10 +946,14 @@ def _find_latest_checkpoint(output_dir):
                     continue
 
                 checkpoint_path = os.path.join(output_dir, name)
-                incomplete_marker = os.path.join(checkpoint_path, CHECKPOINT_INCOMPLETE_MARKER)
+
+                # Check for any incomplete marker files (supports per-rank markers)
+                has_incomplete = any(
+                    f.startswith(CHECKPOINT_INCOMPLETE_MARKER) for f in os.listdir(checkpoint_path)
+                )
```
Comment (hrathina) [inline kubeflow/trainer/rhai/transformers.py]: `os.listdir(checkpoint_path)` can crash with `FileNotFoundError` when rank 0 deletes the checkpoint while other ranks are reading it. Wrap in try-except to gracefully skip checkpoints deleted during concurrent access.

```suggestion
         try:
              files = os.listdir(checkpoint_path)
              has_incomplete = any(f.startswith(CHECKPOINT_INCOMPLETE_MARKER) for f in files)
        except (FileNotFoundError, OSError):
              # Rank 0 deleted this checkpoint while we were reading
              _log(f"Checkpoint {name} deleted by another rank, skipping")
              continue
```

File: kubeflow/trainer/rhai/transformers.py [hunk 6]
```diff
@@ -258,55 +276,62 @@ def setup_signal_handler(self):
             _log("JIT checkpoint signal handler registered for SIGTERM")
 
         def _sigterm_handler(self, signum, frame):
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py:277]: how does training actually stop once the sigterm handler exits? Do we have to wait for the training step to end? Might that trigger another checkpoint?

Comment (efazal) [reply kubeflow/trainer/rhai/transformers.py:277]: No we don't wait until the training step completes, on_pre_optimizer stage the flag `_in_optimizer_step` gets set to True, and in `on_optimizer_step` callback we begin jit checkpoint. If you are wondering, when the JIt checkpoint get deffered would this trigger a periodic checkpoint before JIT checkpoint? no, it won't, since we mark `should_save` flag as false, which decides if a periodic checkpoint should be taken, and also we set `should_training_stop` to true , which ensures the training halts.

Comment (robert-bell) [reply kubeflow/trainer/rhai/transformers.py:277]: What about if we're not in the optimiser step? The JIT checkpoint is triggered immediately, but once it's completed how does the training loop terminate? Do we have to wait until the step completes?

Comment (efazal) [reply kubeflow/trainer/rhai/transformers.py:277]: In the signal handler we set `checkpoint_requested` to true, so JIT checkpoint completes, it will hit at least one of the callbacks like `on_pre_optimizer_step` or `on_step_end` which will again set `should_training_stop` to True. And it will ensure the training stops.

Comment (robert-bell) [reply kubeflow/trainer/rhai/transformers.py:277]: so we do need to wait for the step to complete? How long does that take? Could this be multiple hours?

Comment (robert-bell) [reply kubeflow/trainer/rhai/transformers.py:277]: Discussed offline and we've agreed the implementation is ok.

- For llama-70b with PEFT, it's ~ 10mins per step; without PEFT maybe an hour. However, the JIT checkpoint will be uploaded to s3 immediately in the background, so provided the graceful shutdown period is long enough the JIT checkpoint will appear in s3.
- there's a small risk the sigterm occurs during an existing checkpoint. The handler may take another checkpoint on top of the one in progress, but any files will just get overwritten. The likelihood of this happening is low, and the consequences are negligible so we're going to leave the implementation as is.

File: kubeflow/trainer/rhai/transformers_test.py [hunk 7]
```diff
@@ -1979,13 +1979,15 @@ def test_find_latest_checkpoint(test_case, tmp_path):
         checkpoint_path.mkdir()
 
         if checkpoint_config.get("incomplete"):
-            (checkpoint_path / CHECKPOINT_INCOMPLETE_MARKER).write_text("incomplete")
+            # Use per-rank marker format matching production code
+            marker_name = f"{CHECKPOINT_INCOMPLETE_MARKER}.node-0-rank-0"
+            (checkpoint_path / marker_name).write_text("incomplete")
```
Comment (hrathina) [inline kubeflow/trainer/rhai/transformers_test.py:1991]: nit: good to have a test case verifying the local JIT checkpoint sentinel lifecycle: sentinel created before _save_checkpoint(), removed after successful save, and checkpoint is resume-ready with no CHECKPOINT_INCOMPLETE_MARKER files present.

**Comment by ChughShilpa** [comment]
Tested it locally and it is working as expected

**Review by ChughShilpa — APPROVED** [review]
